session/session.go

```diff
@@ -963,29 +963,30 @@ func (s *session) GetGlobalSysVar(name string) (string, error) {
 
 // SetGlobalSysVar implements GlobalVarAccessor.SetGlobalSysVar interface.
 func (s *session) SetGlobalSysVar(name, value string) error {
+	sv := variable.GetSysVar(name)
+	if sv == nil {
+		return variable.ErrUnknownSystemVar.GenWithStackByArgs(name)
+	}
 	if name == variable.TiDBSlowLogMasking {
 		name = variable.TiDBRedactLog
 	}
-	if name == variable.SQLModeVar {
-		value = mysql.FormatSQLModeStr(value)
-		if _, err := mysql.GetSQLMode(value); err != nil {
-			return err
-		}
-	}
 	var sVal string
 	var err error
-	sVal, err = variable.ValidateSetSystemVar(s.sessionVars, name, value, variable.ScopeGlobal)
+	sVal, err = sv.Validate(s.sessionVars, value, variable.ScopeGlobal)
 	if err != nil {
 		return err
 	}
+	if err = sv.SetGlobalFromHook(s.sessionVars, sVal); err != nil {
+		return err
+	}
 	name = strings.ToLower(name)
 	// update mysql.tidb if required.
 	if s.varFromTiDBTable(name) {
 		if err = s.setTiDBTableValue(name, sVal); err != nil {
 			return err
 		}
 	}
-	variable.CheckDeprecationSetSystemVar(s.sessionVars, name, sVal)
+	variable.CheckDeprecationSetSystemVar(s.sessionVars, name)
```

Comment:
I will move this into the sysvar struct in a future PR. It should be possible to declaratively state that a sysVar is deprecated.